### ISSUE TITLE
Make Yum module handle yum locks without hanging.

### ISSUE THREAD

**Issue by gamuniz** [ansible/ansible#44120]
##### SUMMARY

If there is a lock on yum with a similar error:

```
Another app is currently holding the yum lock; waiting for it to exit...
  The other application is: yum
```

##### ISSUE TYPE
 - Feature Idea

##### ANSIBLE VERSION

```
ansible --version
ansible 2.6.2
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/root/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/site-packages/ansible
  executable location = /bin/ansible
  python version = 2.7.5 (default, Nov 20 2015, 02:00:19) [GCC 4.8.5 20150623 (Red Hat 4.8.5-4)]
```

##### OS / ENVIRONMENT

N/A

**Comment by jctanner** [comment]
Situation can be reproduced by running yum in another terminal or by creating a lock file.

```
echo 1 > /var/run/yum.pid
```

**Comment by webknjaz** [comment]
Lock is there for a reason, I'm pretty sure that it's unsafe to automagically remove it.

**Comment by jctanner** [comment]
We're not going to remove it, we're going to detect the lock and fail the module accordingly.

**Comment by jctanner** [comment]
prototype patch:

```diff
diff --git a/lib/ansible/module_utils/yumdnf.py b/lib/ansible/module_utils/yumdnf.py
index 0d53914462..219e82fc4b 100644
--- a/lib/ansible/module_utils/yumdnf.py
+++ b/lib/ansible/module_utils/yumdnf.py
@@ -11,6 +11,7 @@
 
 import os
 import tempfile
+import time
 from abc import ABCMeta, abstractmethod
 
 from ansible.module_utils._text import to_native
@@ -42,6 +43,8 @@ yumdnf_argument_spec = dict(
         update_cache=dict(type='bool', default=False, aliases=['expire-cache']),
         update_only=dict(required=False, default="no", type='bool'),
         validate_certs=dict(type='bool', default=True),
+        yumlock_poll=dict(type='int', default=-1),
+        yumlock_sleep=dict(type='int', default=10)
         # this should not be needed, but exists as a failsafe
     ),
     required_one_of=[['name', 'list']],
@@ -97,3 +100,14 @@ class YumDnf(with_metaclass(ABCMeta, object)):
     @abstractmethod
     def run(self):
         raise NotImplementedError
+
+
+def wait_for_yum_lock(lockfile='/var/run/yum.pid', seconds=10, interval=-1):
+    '''Poll until the lock is removed if interval is a positive number'''
+    if os.path.isfile(lockfile) and interval > 0:
+        for iteration in range(0, interval):
+            time.sleep(seconds)
+            if not os.path.isfile(lockfile):
+                break
+        if os.path.isfile(lockfile):
+            raise Exception('YUM lockfile was not released')
diff --git a/lib/ansible/modules/packaging/os/yum.py b/lib/ansible/modules/packaging/os/yum.py
index 2fe93fb7a0..50b9684192 100644
--- a/lib/ansible/modules/packaging/os/yum.py
+++ b/lib/ansible/modules/packaging/os/yum.py
@@ -309,7 +309,7 @@ EXAMPLES = '''
 from ansible.module_utils.basic import AnsibleModule
 from ansible.module_utils._text import to_native
 from ansible.module_utils.urls import fetch_url
-from ansible.module_utils.yumdnf import YumDnf, yumdnf_argument_spec
+from ansible.module_utils.yumdnf import YumDnf, yumdnf_argument_spec, wait_for_yum_lock
 
 import os
 import re
@@ -1429,6 +1429,11 @@ class YumModule(YumDnf):
         if not HAS_YUM_PYTHON:
             error_msgs.append('The Python 2 yum module is needed for this module. If you require Python 3 support use the `dnf` Ansible module instead.')
 
+        wait_for_yum_lock(
+            seconds=self.module.params['yumlock_sleep'],
+            interval=self.module.params['yumlock_poll']
+        )
+
         if self.disable_excludes and yum.__version_info__ < (3, 4):
             self.module.fail_json(msg="'disable_includes' is available in yum version 3.4 and onwards.")
 
@@ -1506,6 +1511,8 @@ def main():
     )
 
     module_implementation = YumModule(module)
+    for x in dir(module_implementation.module):
+        import q; q(x)
     module_implementation.run()
```
